### PR TITLE
Add option to require principal approval

### PR DESCRIPTION
**Pull request by megcrenshaw**
Modifies the frontend to be able to both require and un-require principal approval.

There's some weirdness here. There are lots of cases, and most of the logic for this lives here: https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/models/pd/application/teacher_application.rb#L423-L436. You can see all the cases down below, but in short, we don't know if a principal approval is required based on `principal_approval_state` alone. So, I split out `principal_approval_state` and `principal_approval_not_required`:
- In the detail view, that looks like only showing the `principal_approval_state` and the `principal_approval_not_required` if the approval is not completed (could be required or not required, and incomplete).
- In the table view, it looks like the current code allows us to only show `principal_approval_state` and the `principal_approval_not_required` separate from each other –– so we can't change one view based on the other. We could add extra state to allow for these values to change (for example, not showing "Make not required" button if the approval is already complete), but I'm not sure that should be prioritized. Instead, I went with having two separate columns for these, rather than only the `principal_approval_not_required` button.

Functionality: https://watch.screencastify.com/v/00qrnAEe7iZ10yHmaZa9 

## Testing story

I added backend tests for the endpoint. For the frontend, I added some tests for the detail view. For the quick view, adding reasonable tests would require a `mount`, and doing that requires a refactor of the component itself to avoid state leaks. So, I stuck with testing in detail view. BUT there are no tests looking at how the frontend behaves with different server responses –– could add those in too if we think that'd be good.

Here are the cases I tested out manually, looking at both the detail view and the table view:

1) Principal approval is required, and is complete. `principal_approval_not_required` will be `nil` and `principal_approval_state` will be `Complete`.
2) Principal approval is required, an email has been sent but is not completed. `principal_approval_not_required` will be `nil` and `principal_approval_state` will be `Incomplete`
3) Principal approval is required, but email has not been sent. Theoretically, this will only happen on `incomplete` teacher applications –– the email gets triggered when it moves from `incomplete` to `unreviewed`. In this state, `principal_approval_not_required` will be `nil` and `principal_approval_state` will be `nil`.

4) Principal approval is not required, is not complete, and no email has ever been sent. `principal_approval_not_required` will be `nil` and `principal_approval_state` will be `Not required`
5) Principal approval is not required, is not complete, but an email was sent. This would happen if, after an email has already been sent, we mark the principal approval as not required. `principal_approval_not_required` will be `nil` but `principal_approval_state` will be `Incomplete` –– we can't rely on `principal_approval_state` to accurately tell us about `principal_approval_not_required`
6) Principal approval is not required, but the approval has been completed. This would happen if, after an email has already been sent, we mark the principal approval as not required. `principal_approval_not_required` will be `nil` but `principal_approval_state` will be `Complete` –– here, too, we can't rely on `principal_approval_state` to accurately tell us about `principal_approval_not_required`

## Follow-up work

I do think if we are wanting to make more changes to the quick_view_table, we should refactor the component using more updated patterns (e.g. a function component with hooks for state). This will also make it easier to test. Not sure the best way to document this.

Work item to remove dashboard_db_access from UI test: https://codedotorg.atlassian.net/browse/PLAT-1951

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js
Comment (megcrenshaw) [inline apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:210]: These tests are unchanged, but they were also tested with facilitator applications, which wasn't necessary

File: apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js
Comment (megcrenshaw) [inline apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:416]: I went back and forth on how thoroughly to test these. I could ensure that once this button is clicked, the right things happen –– the right data gets sent, we set state correctly. But testing state changes isn't good (instead, I'd test that the new button shows up, from Make required to Make not required or vice-versa), and we aren't testing other ajax calls in this test suite. But happy for pushback here –– I went with the easiest, which isn't always the best ;)

File: dashboard/app/models/pd/application/teacher_application.rb
Comment (megcrenshaw) [inline dashboard/app/models/pd/application/teacher_application.rb:265]: This was being overwritten in a different principal_approval_state method 🤯 ...

Comment (davidsbailey) [reply dashboard/app/models/pd/application/teacher_application.rb:265]: nice catch! can you help me understand the other method a bit better? specifically, why does the existence of a principal approval application indicate that approval is complete? naively, it seems like completeness should depend on the contents of `do_you_approve`: https://github.com/code-dot-org/code-dot-org/blob/3c581f43c3e38a0ca293d60d7bbac0b471206169/dashboard/app/models/pd/application/teacher_application.rb#L423-L425

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx
Comment (davidsbailey) [inline apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:990]: for future consideration: it would be better if application behavior did not depend on UI string contents. this approach is fragile in case we (1) change the strings or (2) start to translate them. my recommendation would be to use two separate variables, one to control application behavior ('incomplete', 'complete', etc) and another containing the UI string.

File: apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js
Comment (davidsbailey) [reply apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:416]: all makes sense. my understanding is that this is an area where we have decided that we do want to invest in better test coverage in code we are touching. Based on that, I do think it would be worth making adding a bit more test coverage now. this could either be the kind of test you landed on above, or UI tests (they would look a lot like your screen recordings). since the principal approval button appears in a couple different places, I think the better bang-for-the-buck would be the kind of test you were talking about:

> test that the new button shows up, from Make required to Make not required or vice-versa

there should be examples of other tests doing something similar using sinon to mock the ajax requests to provide server responses. let me know if you want help identifying a good example to follow.

**Review by davidsbailey — COMMENTED** [review]
nice progress Meg! Thank you for the detailed PR writeup. Unfortunately I think I am having trouble understanding some of the weirdnesses you are talking about there. I grabbed some time tomorrow, maybe you can give me a quick demo?

File: apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js
Comment (davidsbailey) [reply apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:416]: quick follow-up: I guess you would still need a separate test for each view regardless of whether you are using a js test or a UI test 🤷‍♂️ my suggestion would still be to focus on improving the js tests to cover the new behavior. UI tests could be a nice bonus.

Comment (megcrenshaw) [reply apps/test/unit/code-studio/pd/application_dashboard/detail_view_contentsTest.js:416]: Sounds good –– I think mocking the ajax requests makes sense. I did that for the teacher apps, so I'm guessing it'll look similar.

File: dashboard/app/models/pd/application/teacher_application.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/teacher_application.rb:265]: A principal approval is only created when it's submitted –– the link to it doesn't create one on the backend. So for example, on staging, there's a link to a principal approval here https://staging-studio.code.org/pd/application_dashboard/csa_teachers/11 but the record doesn't exist until it's been submitted.

So "Complete" only refers to the principal approval being submitted (i.e. completed), but the approval itself depends on the :do_you_approve part

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx
Comment (megcrenshaw) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:990]: Totally agree. The `principal_approval_state` uses the COMPLETE constant –– should I import that in?

**Comment by davidsbailey** [comment]
All sounds good Meg! I'd like to take another look once you've got those tests added.

File: apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx
Comment (davidsbailey) [reply apps/src/code-studio/pd/application_dashboard/detail_view_contents.jsx:990]: oh, interesting. yes, comparing for equality against that constant seems like a nice improvement!

EDIT: that would protect against (1), and (2) is probably a long way off.

**Comment by megcrenshaw** [comment]
Decided to add a UI test to have coverage in both DetailView and QuickView ... extracted the principal approval states as constants too!

File: dashboard/app/models/pd/application/teacher_application.rb
Comment (davidsbailey) [inline dashboard/app/models/pd/application/teacher_application.rb:413]: very creative!

File: dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature
Comment (davidsbailey) [inline dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature:1]: this can be slow and therefore flaky. can you open a follow-up item to eliminate this, and we can discuss at midsprint? the work would be to define an api in test_controller.rb, and then hit that api to do the necessary setup rather than access the dashboard db directly. I am guessing it is just this that needs it: https://github.com/code-dot-org/code-dot-org/blob/46b26f2d5eb43dfb02fc3cb4416a14c34e62d76d/dashboard/test/ui/features/step_definitions/pd.rb#L1-L9

**Review by davidsbailey — APPROVED** [review]
nice improvements, nice test!

File: dashboard/app/models/pd/application/teacher_application.rb
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/teacher_application.rb:413]: This was an effect of how the constants get exported from ruby to jsx. I needed to change the constant from an array to an object, which meant having a less-nice approach for these in ruby ...

File: dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature
Comment (megcrenshaw) [reply dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature:1]: Ah, I see. Yeah, I used that because that's what I saw in `teacher_application_dashboard.feature` and I figured it needed it to create fake applications.

I'm most worried about this interacting with `teacher_application_dashboard.feature` –– I only need one application, but don't want to create a new one that could mess up `teacher_application_dashboard.feature`. The `creating some fake applications of each type and status` step doesn't create new apps if they already exist:

https://github.com/code-dot-org/code-dot-org/blob/9b49f259b8f1b5646dd4a583b7efd457dc45ccaf/dashboard/test/ui/features/step_definitions/pd.rb#L162-L165

Anyways, we can chat at midsprint. Shall I wait to merge this, or create a separate work item to address?

Comment (davidsbailey) [reply dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature:1]: separate work item for eliminating `@dashboard_db_access` is fine. if you are asking about holding off on merging because of potential flakiness due to two UI tests both creating an application, let's try to evaluate that concern before you merge.

Comment (megcrenshaw) [reply dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature:1]: Good call. I think we could run into trouble if (a) we need to create an application and (b) these tests are running at the exact same time -- that could create two apps of the same kind

Comment (davidsbailey) [reply dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature:1]: good point about tests running simultaneously. to that point, simply cleaning up at the beginning of each scenario wouldn't work well.

I won't have time to dig in til after demos, but here are a couple half-baked ideas:
1. the `@only_one_browser` flag, though this would still mean you have two test runs (one eyes, one non)
2. can you identify the application, and find it in the browser window, by something other than assuming it is the first and only application?

Comment (megcrenshaw) [reply dashboard/test/ui/features/ed_programs/pd/application_detail_view.feature:1]: Just checked the `teacher_application_dashboard` test –– we're ignoring the regions that show how many apps are created, and it doesn't matter which one gets clicked, so long as it has the right status. So, I think we're safe here

**Comment by megcrenshaw** [comment]
Follow-up work item to remove dashboard_db_access from UI test: https://codedotorg.atlassian.net/browse/PLAT-1951

**Comment by davidsbailey** [comment]
all looks good!

**Pull request merged**
